src/Reflex/Dom/Contrib/Router.hs

```diff
@@ -3,81 +3,191 @@
-                (leftmost $ map _routeConfig_pushState rcs)
 
-data Route t = Route {
-    _route_value :: Dynamic t Text -- ^ URL value
+data Route t a = Route {
+    _route_value   :: Dynamic t (Either T.Text a)     -- ^ Routing value
+  , _route_fullURI :: Dynamic t (U.URIRef U.Absolute) -- ^ Full URI
   }
 
-instance HasValue (Route t) where
-  type Value (Route t) = Dynamic t Text
+makeLenses ''Route
+
+instance HasValue (Route t a) where
+  type Value (Route t a) = Dynamic t (Either T.Text a)
   value = _route_value
 
--- | Manipulate and track the URL text for dynamic routing of a widget
-route :: (HasWebView m, MonadWidget t m) => RouteConfig t -> m (Route t)
-#if ghcjs_HOST_OS
-route (RouteConfig goForward goBack sSet) = do
-  win <- askDomWindow
-  loc <- getLocation' win
+-------------------------------------------------------------------------------
+-- | Manipulate and track the URL 'GHCJS.DOM.Types.Location' for dynamic
+--   routing of a widget
+route
+  :: (HasWebView m, MonadWidget t m)
+  => (Location -> IO (Either T.Text a))
+     -- ^ Decode the part of the path beyond '_routeConfig_pathBase' into an a
+  -> (a -> IO T.Text)
+     -- ^ Encode the routing value to a Text that will be passed to `pushState`
+  -> RouteConfig t a
+    -- ^ Routing widget configuration
+  -> m (Route t a)
+route to from (RouteConfig goForward goBack sSet) = do
+  win     <- askDomWindow
+  loc0    <- liftIO $ windowURI win
+  locVal0 <- liftIO $ to =<< fmap (fromMaybe (error "No location"))
+                                  (getLocation win)
+
   Just hist <- liftIO $ getHistory win
   performEvent_ $ ffor goForward $ \_ -> liftIO (forward hist)
   performEvent_ $ ffor goBack    $ \_ -> liftIO (back hist)
-  setLoc <- performEvent $ ffor sSet $ \t -> do
-    pushState hist (pToJSVal (0 :: Int)) ("" :: Text) t
-    getLocation' win
-  newLocs <- getPopState
-  Route <$> holdDyn loc (leftmost [setLoc, newLocs])
-#else
-route = error "route is only available to ghcjs"
-#endif
+
+  _ <- performEvent $ ffor sSet $ \t -> liftIO $ do
+    s <- from t
+    pushState hist (pToJSVal (0 :: Int)) ("" :: T.Text) s
+    dispatchEvent'
+
+  locVals <- getPopState to
+  locs <- performEvent $ ffor locVals $ \_ -> liftIO (windowURI win)
+
+  Route <$> holdDyn locVal0 (leftmost [Right <$> sSet, locVals]) <*> holdDyn loc0 locs
+
+
+windowURI :: Window -> IO (U.URIRef U.Absolute)
+windowURI w = do
+  l <- fromMaybe (error "Window has no location") <$> getLocation w
+  either (error "No parse of window location") id . U.parseURI U.laxURIParserOptions . T.encodeUtf8 <$> toString l
+
+-------------------------------------------------------------------------------
+-- | Route a single page app according to any 'WR.PathInfo' a => a
+webRoute
+  :: (MonadWidget t m, WR.PathInfo a)
+  => T.Text     -- ^ The part of the URL not related to SPA routing
+  -> RouteConfig t a
+  -> m (Route t a)
+webRoute pathBase = route decoder (return . (pathBase <>) . WR.toPathInfo)
```

Comment:
In that case, I think I'll just make `route`'s api use URIRef's directly. After getting rid of web-route and looking at the way the de/serialization parameters to `route` are currently being used, I don't see any reason to make `route` polymorphic. `fullUriRoute` and `webRoute` could just as well have been written as a wrappers around a monomorphic version of `route`, rather than as a special case of a polymorphic `route`. Any objection?